### ISSUE TITLE
[New rule] `tag_name` Log tag is the same as class name

### ISSUE THREAD

**Issue by fzyzcjy** [dart-code-checker/dart-code-metrics#713]
**Please describe what the rule should do:**
Warn when the tag (such as `static const _kTag = '...'`) is different from the name of the class containing the tag.
Tags are used frequently in logging, and such outdated/typo tag can cause confusion when reading the log.

**If your rule is inspired by other please provide link to it:**

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

```dart
class OneService {
  static const _kTag = 'OneService';
}

class OneService {
  static const _kTag = 'TwoService'; // LINT
}
```

**Are you willing to submit a pull request to implement this rule?**

**Comment by fzyzcjy** [comment]
https://github.com/dart-code-checker/dart-code-metrics/pull/715

**Comment by incendial** [comment]
> Tags are used frequently in logging, and such outdated/typo tag can cause confusion when reading the log.

Could you share some public code that uses this approach? Never saw it before, just curious, how popular it is.

**Comment by fzyzcjy** [comment]
> Could you share some public code that uses this approach? Never saw it before, just curious, how popular it is.

Firstly, a private codebase with >150kloc: My code :)
Secondly, in Android logging it is the standard approach: https://developer.android.com/reference/android/util/Log Officially it says:

> Tip: A good convention is to declare a TAG constant in your class:
private static final String TAG = "MyActivity";
and use that in subsequent calls to the log methods.

Thirdly, I am also curious, how do you usually do logging, without such approach? Or how do public code often do logging? I can do refactoring if something better is the de-facto standard. @incendial

**Comment by incendial** [comment]
> > Could you share some public code that uses this approach? Never saw it before, just curious, how popular it is.
> 
> Firstly, a private codebase with >150kloc: My code :) Secondly, in Android logging it is the standard approach: https://developer.android.com/reference/android/util/Log Officially it says:
> 
> > Tip: A good convention is to declare a TAG constant in your class:
> > private static final String TAG = "MyActivity";
> > and use that in subsequent calls to the log methods.
> 
> Thirdly, I am also curious, how do you usually do logging, without such approach? Or how do public code often do logging? I will plan to do refactoring if something better is the de-facto standard. @incendial

I'm not working with Flutter for a past few month (coz of a lot of work), but usually I use https://pub.dev/packages/logging with a string class name. I asked because I never saw `TAG` naming before.

**Comment by fzyzcjy** [comment]
> I asked because I never saw TAG naming before.

I came from Android (~10 years ago I developed android) as well as Java so I am familiar with it :)

> with a string class name

Well I guess the problem is, if you write `log('MyClassName', ...)`, say, 10 times inside that class, it may violate the DRY rule.

**Comment by incendial** [comment]
> Well I guess the problem is, if you write log('MyClassName', ...), say, 10 times inside that class, it may violate the DRY rule.

Agree, but usually it's a logger per class and they are nested, so the name is not repeated.

**Comment by fzyzcjy** [comment]
@incendial I see. So you are creating logger objects instead of using static methods.

**Comment by dkrutskikh** [comment]
Available in 4.12.0 release 🚀
